src/grpc/stream_infer_handler.cc

```diff
@@ -540,57 +544,104 @@ ModelStreamInferHandler::StreamInferResponseComplete(
     }
 
     if (err != nullptr) {
       ::grpc::Status status;
       GrpcStatusUtil::Create(&status, err);
       response->mutable_infer_response()->Clear();
       response->set_error_message(status.error_message());
-
-      // repopulate the id so that client knows which request failed.
-      const char* id;
-      LOG_TRITONSERVER_ERROR(
-          TRITONSERVER_InferenceResponseId(iresponse, &id),
-          "couldn't retrieve id for failed request");
-      LOG_VERBOSE(1) << "Failed for ID: " << id << std::endl;
-      response->mutable_infer_response()->set_id(id);
+      LOG_VERBOSE(1) << "Failed for ID: " << request_id << std::endl;
     }
 
     TRITONSERVER_ErrorDelete(err);
-
     LOG_TRITONSERVER_ERROR(
         TRITONSERVER_InferenceResponseDelete(iresponse),
         "deleting GRPC inference response");
   }
 
-  state->complete_ = ((flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) != 0);
-  if (!state->is_decoupled_) {
-    state->step_ = Steps::WRITEREADY;
-    state->context_->WriteResponseIfReady(state);
-  } else {
+  // Decoupled backends can return a null response via
+  // TRITONBACKEND_ResponseFactorySendFlags. By default, these null
+  // "empty" responses are not sent back to the client. Clients can
+  // opt-in to receiving these empty responses via request parameters.
+  // NOTE: The complete flag is the only flag used for this case at this time.
+  bool empty = (!iresponse && state->is_decoupled_ && state->complete_);
+  bool enable_empty = false;
+  if (empty) {
+    // Check request for opt-in parameter
+    const auto& request_params = state->request_.parameters();
+    const auto iter = request_params.find("triton_enable_empty_response");
+    if (iter != request_params.end()) {
+      const auto& param = iter->second;
+      if (param.parameter_choice_case() ==
+          inference::InferParameter::ParameterChoiceCase::kBoolParam) {
+        if (param.bool_param()) {
+          enable_empty = true;
+        }
+      } else {
+        LOG_ERROR << "invalid value type for "
+                     "'triton_enable_empty_response' parameter, expected "
+                     "bool_param.";
+      }
+    }
+  }
+
+  // Send empty response when detected and opted-in by client/request
+  const bool send_empty_response = (empty && enable_empty);
+  if (send_empty_response) {
+    // Create a response to send. Assume decoupled here based on prior checks.
+    state->response_queue_->AllocateResponse();
+    response = state->response_queue_->GetLastAllocatedResponse();
+    if (response) {
+      LOG_VERBOSE(1) << "[request id: " << request_id << "] "
+                     << "Sending flags-only response";
+      response->mutable_infer_response()->Clear();
+    } else {
+      LOG_ERROR << "expected the response allocator to have added the response";
+    }
+  }
+
+  // Set response parameters to indicate certain behaviors to client
+  if (response) {
+    response->mutable_infer_response()->set_id(request_id);
```

Comment:
If I follow correctly above in the error case we would pass along an empty request_id - but in this case we update it to "<<id unknown>>"

I think we would want to keep it unset. For logging I would potentially update to <<id not set>> or <<id empty>> instead of unknown - as unknown implies its set but not known ....